File: app/src/main/java/com/mohnage7/movies/repository/MoviesRepository.java
package com.mohnage7.movies.repository;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.mohnage7.movies.api.ApiResponse;
import com.mohnage7.movies.api.RestApiService;
import com.mohnage7.movies.base.DataWrapper;
import com.mohnage7.movies.db.AppExecutors;
import com.mohnage7.movies.db.MovieDao;
import com.mohnage7.movies.db.MovieDatabase;
import com.mohnage7.movies.model.Movie;
import com.mohnage7.movies.model.MoviesResponse;
import com.mohnage7.movies.utils.RefreshRateLimiter;

import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.mohnage7.movies.utils.Constants.CACHE_TIMEOUT;

public class MoviesRepository {

    private final AppExecutors appExecutors;
    private RestApiService apiService;
    private MutableLiveData<DataWrapper<List<Movie>>> mutableSearchMovieLiveData = new MutableLiveData<>();
    private MovieDao movieDao;
    private RefreshRateLimiter refreshRateLimiter;

    @Inject
    public MoviesRepository(RestApiService apiService, MovieDatabase movieDatabase, AppExecutors appExecutors) {
        this.apiService = apiService;
        this.appExecutors = appExecutors;
        movieDao = movieDatabase.getMovieDao();
        refreshRateLimiter = new RefreshRateLimiter(TimeUnit.MINUTES, CACHE_TIMEOUT);
    }

    public LiveData<DataWrapper<List<Movie>>> getMovies(String category) {
        return new NetworkBoundResource<List<Movie>, MoviesResponse>(appExecutors) {

            @Override
            protected void saveCallResult(@NonNull MoviesResponse item) {
                addFilterToEveryMovie(item.getMovieList(), category);
                movieDao.insertAll(item.getMovieList());
            }

            @Override
            protected boolean shouldFetch(@Nullable List<Movie> data) {
                return data == null || data.isEmpty() || refreshRateLimiter.shouldFetch(category);
            }

            @NonNull
            @Override
            protected LiveData<List<Movie>> loadFromDb() {
                return movieDao.getAllMovies(category);
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MoviesResponse>> createCall() {
                return apiService.getMovies(category);
            }
        }.getAsLiveData();
    }

    private void addFilterToEveryMovie(List<Movie> movieList, String category) {
        for (Movie movie : movieList) {
            movie.setCategory(category);
        }
    }


    public LiveData<DataWrapper<List<Movie>>> search(String query) {
        mutableSearchMovieLiveData.setValue(new DataWrapper<>(DataWrapper.Status.LOADING, null, null));
        Call<MoviesResponse> call = apiService.search(query);
        call.enqueue(new Callback<MoviesResponse>() {
            @Override
            public void onResponse(@NonNull Call<MoviesResponse> call, @NonNull Response<MoviesResponse> response) {
                MoviesResponse mResponse = response.body();
                if (mResponse != null) {
                    if (mResponse.getMovieList() != null && !mResponse.getMovieList().isEmpty()) {
                        mutableSearchMovieLiveData.setValue(new DataWrapper<>(DataWrapper.Status.SUCCESS, mResponse.getMovieList(), null));
                    } else {
                        mutableSearchMovieLiveData.setValue(new DataWrapper<>(DataWrapper.Status.ERROR, null, mResponse.getStatusMessage()));
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<MoviesResponse> call, @NonNull Throwable error) {
                mutableSearchMovieLiveData.setValue(new DataWrapper<>(DataWrapper.Status.ERROR, null, error.getMessage().equals("") ? error.getMessage() : "Unknown error\nCheck network connection"));
            }
        });
        return mutableSearchMovieLiveData;
    }
}

File: app/src/main/java/com/mohnage7/movies/features/movies/repository/MoviesRepository.kt
package com.mohnage7.movies.features.movies.repository

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.mohnage7.movies.db.AppExecutors
import com.mohnage7.movies.db.CACHE_TIMEOUT
import com.mohnage7.movies.db.RefreshRateLimiter
import com.mohnage7.movies.db.dao.MovieDao
import com.mohnage7.movies.features.movies.model.Movie
import com.mohnage7.movies.features.movies.model.MoviesResponse
import com.mohnage7.movies.network.NetworkBoundResource
import com.mohnage7.movies.network.RestApiService
import com.mohnage7.movies.network.model.ApiResponse
import com.mohnage7.movies.network.model.DataWrapper
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.concurrent.TimeUnit

class MoviesRepository
constructor(private val apiService: RestApiService,
            private val appExecutors: AppExecutors,
            private val movieDao: MovieDao) {

    private val mutableSearchMovieLiveData = MutableLiveData<DataWrapper<List<Movie>>>()
    private val refreshRateLimiter: RefreshRateLimiter<String> = RefreshRateLimiter(TimeUnit.MINUTES, CACHE_TIMEOUT.toLong())

    fun getMovies(category: String): LiveData<DataWrapper<List<Movie>>> {
        return object : NetworkBoundResource<List<Movie>, MoviesResponse>(appExecutors) {

            override fun saveCallResult(item: MoviesResponse) {
                addFilterToEveryMovie(item.movieList!!, category)
                movieDao.insertAll(item.movieList!!)
            }

            override fun shouldFetch(data: List<Movie>?): Boolean {
                return data == null || data.isEmpty() || refreshRateLimiter.shouldFetch(category)
            }

            override fun loadFromDb(): LiveData<List<Movie>> {
                return movieDao.getAllMovies(category)
            }

            override fun createCall(): LiveData<ApiResponse<MoviesResponse>> {
                return apiService.getMovies(category)
            }
        }.asLiveData
    }

    private fun addFilterToEveryMovie(movieList: List<Movie>, category: String) {
        for (movie in movieList) {
            movie.category = category
        }
    }


    fun search(query: String): LiveData<DataWrapper<List<Movie>>> {
        mutableSearchMovieLiveData.value = DataWrapper(DataWrapper.Status.LOADING, null, null)
        val call = apiService.search(query)
        call.enqueue(object : Callback<MoviesResponse> {
            override fun onResponse(call: Call<MoviesResponse>, response: Response<MoviesResponse>) {
                response.body()?.let { mResponse ->
                    if (mResponse.movieList != null && !mResponse.movieList!!.isEmpty()) {
                        mutableSearchMovieLiveData.setValue(
                                DataWrapper(
                                        DataWrapper.Status.SUCCESS,
                                        mResponse.movieList,
                                        null))
                    } else {
                        mutableSearchMovieLiveData.setValue(
                                DataWrapper(
                                        DataWrapper.Status.ERROR,
                                        null, ""))
                    }
                }

            }

            override fun onFailure(call: Call<MoviesResponse>, error: Throwable) {
                mutableSearchMovieLiveData.value = DataWrapper(DataWrapper.Status.ERROR, null, if (error.message == "") error.message else "Unknown error\nCheck network connection")
            }
        })
        return mutableSearchMovieLiveData
    }
}
